### ISSUE TITLE
[Bug] Styles not applied when tab loaded in the background

### ISSUE THREAD

**Issue by xcasxcursex** [darkreader/darkreader#7682]
# Bug Report

## Bug Description
Sites are not styled correctly when loaded in the background

## Website Address
Various sites. Example: https://en.wikipedia.org/wiki/Darkness

## Steps To Reproduce
Load page in the background.

## Expected Behavior
![image](https://user-images.githubusercontent.com/58130821/146639681-dff1780f-79e0-4270-8035-033bb19652e0.png)

## Actual Behavior
I loaded the page and took the above screen shot.
I ctrl+tabbed to this tab to paste the screen shot
I right-click the above tab and select reload, while I am still in this tab
ctrl+tab back to the page and this is what I see:
![image](https://user-images.githubusercontent.com/58130821/146639695-1c086cd5-01c2-4f45-9f09-b516f3e6d92c.png)

## Screenshots
See above

## System Information

- OS: Win10
- Browser: Firefox Nightly 97
- Dark Reader Version:  4.9.43

## Additional Context
This started happening recently,

**Comment by Gusted** [comment]
I cannot reproduce this.

Do you have any other extension that might interfere, also could you check if this reproduces on a new profile(about:profiles)?

Regards,
Gusted

**Comment by msdn77** [comment]
I can confirm the same. When i load a site in current tab Dark Reader styles site as it should be.
But when i load an site in background tab it doesnt do anything expect coloring background and text color.


Example of bug:

In current tab:
![1](https://user-images.githubusercontent.com/65540402/147419492-f9f31360-cb0a-42c2-a6c3-83ff1dfae5f4.jpg)



In background tab:
![2](https://user-images.githubusercontent.com/65540402/147419505-cf920aeb-ebf6-42d3-abad-bbcd770fb7b9.jpg)




No other extensions installed.

OS: Win7
Browser: Firefox 95.0.2
Dark Reader Version: 4.9.43

**Comment by msdn77** [comment]
Also, if "When you open a link, image or media in a new tab, switch to it immediately" option in Firefox is enabled then Dark Reader applies dynamic style correctly.

**Comment by msdn77** [comment]
https://user-images.githubusercontent.com/65540402/147421459-82349742-90ec-41fb-bb46-fd9fbaad3253.mp4

**Comment by msdn77** [comment]
https://user-images.githubusercontent.com/65540402/147421526-9f4c0e75-2cf1-4257-9fc7-76449c5041e4.mp4

**Comment by xcasxcursex** [comment]
Sorry I've been a little slow to report back on this. The behaviour was changing slightly with newer nightly releases, so I wanted to wait it out a little while. It seems to have stabilised now, and the wikipedia link in the OP now only stays un-darkened for a brief moment before it styles correctly...,. but it's still definitely broken, this flicker when changing to the tab is no good.

Another issue I'm having here is that the *most* effected page I can find, is my aliexpress orders page, which I can't really screenshot easily because it's full of private info.

I tried to manually apply the dynamic theme as per the above post, but no luck. Page remains un-styled until I toggle dark reader off and on, or just reload the page in the foreground.

**Comment by Gusted** [comment]
> this flicker when changing to the tab is no good

The flicker is because of how browsers renders the websites, you need to be sure that the OS's dark mode is enabled(and that firefox can detect this) to minimize it, but nothing much to fix on that.


I still cannot reproduce any bugs with background tabs, I've been using them more for the last week in order to check if I can reproduce them, in only 2 instances they where bugged, but not consistent and most likely a flook.

You could try if it's reproduce-able on a new profile(`about:profiles`)

**Comment by xcasxcursex** [comment]
> The flicker is because of how browsers renders the websites, you need to be sure that the OS's dark mode is enabled(and that firefox can detect this) to minimize it, but nothing much to fix on that.

I know the behaviour you mean, but it's not that. What I see is the un-styled page like in my screenshot, but only for a brief moment, then it changes to the styled version. 

I created a brand new profile, added dark reader from darkreader.org, browsed to the link from my post in a new tab, switched back to the dark reader tab, right-clicked the wikipedia tab to reload, clicked the tab when the hourglass was gone plus a couple seconds, and it had the partially-themed thing like in my screenshot, and instantly changed to the themed one. 

I tried 4.9.42 and it's still doing it. Dark reader was definitely not doing this back when that was released, so something must have changed in firefox/nightly I guess.

**Comment by Gusted** [comment]
> What I see is the un-styled page like in my screenshot, but only for a brief moment, then it changes to the styled version.

That's a "hack" from Firefox in order to make it feel that Firefox is more snappier. That's expected behavior and not much dark reader can do.



> and it had the partially-themed thing like in my screenshot, and instantly changed to the themed one.

Wait? It's actually changing to the correct themed one? I thought this issue was about that dark reader didn't apply any or only partial styles to background-loaded tabs.



> so something must have changed in firefox/nightly I guess.

Ah well, they changes a lot of little quirks and behaviors between releases so it's a possibility, but I don't expect them to revert to the old behavior.

**Comment by xcasxcursex** [comment]
> That's a "hack" from Firefox in order to make it feel that Firefox is more snappier. That's expected behavior and not much dark reader can do.

It applies DR's styles before I switch to the tab, but only some of them. After I switch to the tab it applies others. Is that intentional?
See the above screenshots, the image in the top left is inverted (normally 'Wikipedia' is in black text), that's dark reader doing that. I see that inverted image, but with the wrong styling behind it, like in the screenshot, and then it changes to the same image but with the correctly re-styled background behind it.

> > and it had the partially-themed thing like in my screenshot, and instantly changed to the themed one.
> 
> Wait? It's actually changing to the correct themed one? I thought this issue was about that dark reader didn't apply any or only partial styles to background-loaded tabs.

You understood correctly. For example my aliexpress orders page looks like this if I load it in the background and switch to it, and it stays that way.

The wikipedia page *was* doing the same thing, and then it changed and is doing a different thing where it starts out the same then immediately fixes itself.

> Ah well, they changes a lot of little quirks and behaviors between releases so it's a possibility, but I don't expect them to revert to the old behavior.

Exactly, that's why I thought I should let you know that whatever they are changing here in nightly is about to flow down and break sites for your main userbase.

**Comment by Gusted** [comment]
> It applies DR's styles before I switch to the tab, but only some of them. After I switch to the tab it applies others. Is that intentional?
> See the above screenshots, the image in the top left is inverted (normally 'Wikipedia' is in black text), that's dark reader doing that. I see that inverted image, but with the wrong styling behind it, like in the screenshot, and then it changes to the same image but with the correctly re-styled background behind it.

Ah, I'm pretty sure you saw dark reader's fallback style. This style is made to be applied first on all websites(and removed afterwards fairly quickly) it's a simple style that makes the majority of the content dark(to avoid white content flashing someone's eyes) it isn't made to look good which is why the page looks incorrect. Their aren't a lot of situations whereby you can actually notice that this style is applied(without looking very hard for it), this seems to be one of them - which is quite weird.



> You understood correctly. For example my aliexpress orders page looks like this if I load it in the background and switch to it, and it stays that way:

Yeah, that's a real bug, which I should need to reproduce myself in order to get to the bottom of it. I don't have a aliexpress account, do you happen to know a public site whereby you have this behavior?

> Exactly, that's why I thought I should let you know that whatever they are changing here in nightly is about to flow down and break sites for your main userbase.

I'm on nightly myself for this kind of reasons, yet I didn't have this behavior hit me _yet_, It could be that the render engine made a change for windows as I'm currently on linux. But thanks, it's good to note the behavior change.

**Comment by msdn77** [comment]
@Gusted

I posted a video with bug few posts above. It's easily reproducebale on kinozal(dot)tv website.

P.S. sry if I interrupt your conversation.

**Comment by Gusted** [comment]
> I posted a video with bug few posts above. It's easily reproducebale on kinozal(dot)tv website.

Well it's nice that the article's "headers" are not inverted by me(which are on your video). Their are some other global issues with images like that, so it could be that it isn't related to the background tab,

**Comment by msdn77** [comment]
> > I posted a video with bug few posts above. It's easily reproducebale on kinozal(dot)tv website.
> 
> Well it's nice that the article's "headers" are not inverted by me(which are on your video). Their are some other global issues with images like that, so it could be that it isn't related to the background tab,

Headers are processed by the rules in Dev Tools. In my case on kinozal(dot)tv website, i wrote simple rules:

```
kinozal.tv

CSS
.mn_wrap, .mn2, .bx1, .bx2, .bx2_0 {
background-image: url("data:image/svg+xml;base64,"); !important;
}
```

Btw, its something wrong in Firefox engine, at the moment when browser doesnt pass DOM elements to be processed by extension . Because i tested this in Chrome and it works perfectly in 100% cases.

**Comment by xcasxcursex** [comment]
This behaviour is still changing over time. Good news is that it's currently happening on two public sites (instead of needing a login like my previous example):

https://www.guru3d.com/

Foreground reload:
![image](https://user-images.githubusercontent.com/58130821/150959045-c3165c07-2cff-4917-8e68-d47f6e420e84.png)

Background reload:
![image](https://user-images.githubusercontent.com/58130821/150959152-f7f173bc-9e84-456e-89bc-1dab8a69dc8c.png)


https://www.reprap.org/wiki/RepRap_Morgan

Foreground reload:
![image](https://user-images.githubusercontent.com/58130821/150959471-c8b3943c-9426-4779-8280-50d7d68bdaef.png)

Background reload: 
![image](https://user-images.githubusercontent.com/58130821/150959857-345b9f4b-ca50-4b87-b493-17b865557289.png)


It did not happen with a new profile, and did happen with all other addons disabled.

**Comment by xcasxcursex** [comment]
Another example publicly reachable
https://www.head-fi.org/showcase/opera-factory-junior-os1-pro.24990/

Foreground:
![image](https://user-images.githubusercontent.com/58130821/151354509-74c8eb65-e042-406d-b414-82a85709449e.png)

Background:
![image](https://user-images.githubusercontent.com/58130821/151354625-56d37b1e-55af-45cf-b107-70766f1b5ade.png)

**Comment by xcasxcursex** [comment]
Is there anything I can do, to help troubleshoot this? Since I ctrl+click a lot, I'm having to reload a good portion of the pages I view, it's getting a bit tiresome.

**Comment by msdn77** [comment]
> Is there anything I can do, to help troubleshoot this? Since I ctrl+click a lot, I'm having to reload a good portion of the pages I view, it's getting a bit tiresome.

Are you using firefox? try chrome, it still has that issue but in much lower cases

**Comment by xcasxcursex** [comment]
> > Is there anything I can do, to help troubleshoot this? Since I ctrl+click a lot, I'm having to reload a good portion of the pages I view, it's getting a bit tiresome.
> 
> Are you using firefox? try chrome, it still has that issue but in much lower cases

I'll pass on chrome thanks. I'm not trying to find a workaround, I'm trying to find a fix :)


I thought I should try resetting my dark reader settings, which I did, and not only did it not work, now the UI has changed into some new oversimplified thing which doesn't even have the option to import my saved settings!

**Comment by xcasxcursex** [comment]
A detail to help pin this down: I've noticed that, tabs which are in the saved session, can be reloaded in the background, successfully, one time. Newly created tabs fail the first time, all tabs fail subsequent attempts, but those saved session tabs will load with correct styling once. Weird, huh.

**Comment by Gusted** [comment]
> Can this behaviour be disabled, so it will just fully style the page immediately, regardless of its visibility?

I've played with this before and some sites do not play well(might it be on how the sites works or buggy dark reader behavior, it's not worth going trough all the trouble atm). You can modify the extension and remove those parts so it will have that behavior but currently it won't be a option or default in dark reader.

**Comment by Gusted** [comment]
> From Dark Reader's perspective, what is needed to trigger/allow the style to be applied?

Well, in big terms, a few conditions must be met before we actually do the big work, to avoid any issues with making the CPU unhappy when it's just loading in a tab in the background.

- Document.hidden must not be true is it hidden? Listen for visibilitychange event.
- Document.head must be a element, if non-existent? Have mutation observer for it to listen.

Other then that, it shouldn't be blocked by anything in terms of document' events.

**Comment by xcasxcursex** [comment]
So, I measured the delay between switching to a tab and the start of actually re-painting it with the new style. It's 800-900ms. It's like that in a brand-new profile too, and it actually has this delay always, it's just that with foreground tabs the delay is masked by the page layout.

I don't feel like that's not worth the trouble, I feel like that's not working as intended.... but what do I know, I'm not the author ;)

**Comment by Gusted** [comment]
> So, I measured the delay between switching to a tab and the start of actually re-painting it with the new style. It's 800-900ms

Do you have dark mode enabled and `privacy.resistFingerprinting`(in `about:config`) set to false? Because we're checking some things beforehand and we're waiting on the background script to get information about the current tab, so we know which settings to apply etc. If we detect dark mode was enabled on the user, it will apply a style ASAP regardless of any setting.




> I don't feel like that's not worth the trouble, I feel like that's not working as intended.... but what do I know, I'm not the author ;)

It's a very special situation you're in, which not worth the trouble, I meant to refactor the visibility check away and make all the appropriate changes for dark reader in order to support that, but on the other hand, most people don't want dark reader using a lot of CPU when a tab isn't even visible yet.

**Comment by xcasxcursex** [comment]
> Do you have dark mode enabled and `privacy.resistFingerprinting`(in `about:config`) set to false?

I do.

> most people don't want dark reader using a lot of CPU when a tab isn't even visible yet.

I would argue that when the lag is invisible would be the *best* time for it to happen.... Doing it the current way introduces an order of magnitude (! seriously man it's a LOT) increase in latency for everybody, and only holds benefit for devices which don't have enough CPU to load a single page in the background.............. but I'm OK with working around it. 

The trouble is that for DR, if I block the visibilitychange events, that stops DR from changing from the fallback style, but if I fire the visibilitychange event, even with document.hidden forced false and .visibilityState "visible" and the DOM fully loaded and head existing, it doesn't change it. 

While looking at this I've notice that even worse is that it seems that the delay for DR is actually delaying all styling of the page. Here's some code to demonstrate it to you. I'm using violentmonkey but I'm sure you can find a way that suits you.

```

// ==UserScript==
// @name         Hide Visibility
// @namespace    xcasxcursex
// @copyright    xcasxcursex
// @version      0.0.2
// @description  Vis NOYB. Just assume I'm looking at it.
// @author       xcasxcursex
// @match        https://*/*
// @grant        none
// @run-at       document-start
// ==/UserScript==
"use strict";
 
// Try to trick the site into thinking it's never hidden
Object.defineProperty(document, 'hidden', {value: false, writable: false});
Object.defineProperty(document, 'visibilityState', {value: 'visible', writable: false});
Object.defineProperty(document, 'webkitVisibilityState', {value: 'visible', writable: false});
document.dispatchEvent(new Event('visibilitychange'));
document.hasFocus = function () { return true; };

document.addEventListener('visibilitychange', function(e) {
	e.stopImmediatePropagation(); // comment this line out for testing
  testflag();
}, true, true);
 
function testflag() {
  document.getElementsByTagName("body")[0].style.backgroundColor = 'red';
  console.log("hi");
}


```

With this code injected, I can load a tab in the background, and fire the visibilitychange events manually from the console. I see the text feedback and the style of the body element is changed, and clearly the event is stopped since DR doesn't style the page... But the red background never shows up either. Disable DR, and it works as expected - firing the visibilitychange event from the console causes the background to turn red.

What's more, we can use this to see the delay I'm talking about. Comment out the stopImmediatePropagation line, load the tab in the background, fire the event from the console, switch to the tab, and we see the fallback style for a long time before the red is applied. 
Now disable DR and repeat this, and we can see that the red is applied instantly, there is no flash of the white background. Any time after we fire the event, we can see that the background color has been set to red, but it won't render that way until after DR gets its visibilitychange event - unless DR is disabled, then it all just goes.

So, DR is not only saving the lag for when it is visible, it is also making other styling wait until DR replaces its style (when the lag is visible)

I'm pretty sure that's not intentional, right?

Sorry man. I intended to help you avoid some future nightmare with nightly FF and instead I stumbled upon a hornets nest.

**Comment by Gusted** [comment]
> I would argue that when the lag is invisible would be the _best_ time for it to happen.... Doing it the current way introduces an order of magnitude (! seriously man it's a LOT) increase in latency for everybody, and only holds benefit for devices which don't have enough CPU to load a single page in the background.............. but I'm OK with working around it.

I'm very skeptical of removing this "condition", as I can already sniff issues rolling in with all kind of issues that it could cause. For what's worth, on some of the bigger sites, styles are only added after the tab is visible(to prevent your CPU not liking you when you open multiple background tabs), dark reader then again would do the heavy work over those styles, the "initial" code is pretty fast and performant in comparison to the actual modification of styles.



> While looking at this I've notice that even worse is that it seems that the delay for DR is actually delaying all styling of the page. Here's some code to demonstrate it to you. I'm using violentmonkey but I'm sure you can find a way that suits you.

I'm not sure what I should see? `e.stopImmediatePropagation();` is pretty problematic as that stops the dark reader listener from ever receiving notice about the event being executed and removing that LOC would make the page just look as normal + the modified red background color on `body`. All while Dark reader is enabled.




> What's more, we can use this to see the delay I'm talking about. Comment out the stopImmediatePropagation line, load the tab in the background, fire the event from the console, switch to the tab, and we see the fallback style for a long time before the red is applied.

Dark reader says that the document.hidden is true, that's most likely the problem(this is firing the event from console when it's not the active tab):

![image](https://user-images.githubusercontent.com/25481501/154360830-1557469b-39a1-4841-8906-fdab969431dc.png)


> So, DR is not only saving the lag for when it is visible, it is also making other styling wait until DR replaces its style (when the lag is visible)
> 
> I'm pretty sure that's not intentional, right?
> 
> Sorry man. I intended to help you avoid some future nightmare with nightly FF and instead I stumbled upon a hornets nest.

There might be something going on here that shouldn't, but these conditions are pretty carefully chosen in order to avoid particular bugs while the website is loading, which can cause enough troubles on it's own. 5r

**Comment by xcasxcursex** [comment]
> I'm very skeptical of removing this "condition", as I can already sniff issues rolling in with all kind of issues that it could cause. 

FWIW I agree 100%. I think that the way it works now is the only sane default (for reasons of compatibility). But a "immediate background tab processing" option would be nice :) Maybe tucked away in the 'Font & more' menu, or something like that?

> > While looking at this I've notice that even worse is that it seems that the delay for DR is actually delaying all styling of the page. Here's some code to demonstrate it to you. I'm using violentmonkey but I'm sure you can find a way that suits you.
> I'm not sure what I should see? `e.stopImmediatePropagation();` is pretty problematic as that stops the dark reader listener from ever receiving notice about the event being executed 

That is the intention here. We should be able to fire the event, and DR will do nothing because it doesn't get the event, but our script should turn the background red. But it doesn't. When DR doesn't work, no other styling works either.

> and removing that LOC would make the page just look as normal + the modified red background color on `body`.

Right, it should, but it doesn't. The red background only happens when DR does its magic. Unless DR is disabled, then it works as expected and we get the red background immediately. 

That is to say, the object's background-color property is set to 'red', but the actual thing isn't made red on-screen. I just enabled that script and background-reloaded some DDG search results, and switched to the tab to check the background.... here's what's on-screen:
![image](https://user-images.githubusercontent.com/58130821/154390728-d227194c-f289-42aa-9389-03487c1b1a5b.png)

Let's say I type in `document.body.style.backgroundColor = "blue"` at this point, the background still doesn't change.


> Dark reader says that the document.hidden is true, that's most likely the problem(this is firing the event from console when it's not the active tab):

It's set to false and un-writeable by the script, that shouldn't happen. I tested it in the screenshot above.

**Comment by Gusted** [comment]
> FWIW I agree 100%. I think that the way it works now is the only sane default (for reasons of compatibility). But a "immediate background tab processing" option would be nice :) Maybe tucked away in the 'Font & more' menu, or something like that?

That would be possible.

> When DR doesn't work, no other styling works either.

Ah I finally see, the `darkreader-fallback` style is still active(because for all dark reader knows, the tab is currently not being shown) if you manually add `!important;` behind the `blue` inside the DOM editor, you can see it will take effect(as it takes precedence over the fallback style).

**Comment by xcasxcursex** [comment]
I see what you mean about `document.hidden` now. While everything else sees it as false, if I set a breakpoint there, it's true. Strange though, even while paused at that breakpoint if I check the DOM inspector, there it is, false and not writeable. Other scripts tell me false, too.... and `document.visibilityState` behaves the same way. It's strange that only DR sees it this way, but I don't see anything about DR that should cause it, so is this just FF being strange? 

And I also just noticed the !important flags in the fallback style, these two things together explain all of this. 

> That would be possible.

That would be very kind of you. I would work around it if I could, but maybe FF won't allow me to fake the visibility. I'm starting to feel like you would be making a workaround for a problem with FF, but it would also be a nice treat for those times when we have some CPU cycles to spare, and it would make it accessible to DR users without needing to write userscripts, so maybe it would be worthwhile.

**Comment by xcasxcursex** [comment]
FYI, I got curious about this and last night learned a little about developing addons (I know JS already, just needed to know how FF handles it) so I could play with this. 

*Disabling the wait for visibility is AMAZING*. The experience is really, really nice. I ctrl+click a link like normal, switch to the tab, and... no flashes no waiting no wrong colours just glorious dark dynamic mode *instantly*. Zero frame wait time. I'm absolutely certain you'll be widely praised for adding this feature. Anyone who uses background tabs will instantly notice the snappy experience.

**Comment by Gusted** [comment]
> I'm absolutely certain you'll be widely praised for adding this feature.

I'm still skeptic about such change to be included by-default as many will blame and hate dark reader for using a good portion of CPU utilization on a background tab with the reasoning "to just wait until the tab will become active".

**Comment by xcasxcursex** [comment]
Hey, thanks so much mate. 

When I first created this issue I thought it was some new nightly bug we could squash, but in the end I got a privacy boost and a really unexpected but really nice speed boost too! 

Along the way I got to know more about DR and I see you've been very clever in making it easy for not only websites, but also the browser itself and its various addons, to use the DR colours. Hopefully I'll be able to share some patches with other projects, to make their colours 'Dark Reader-aware'. I've already used the vars in my userchrome, it's a really nice DR feature, I didn't know about.


I thought I would take a moment to thank you, and maybe leave some feedback that might be useful for others in the future....

It's really nice to use this new DR setting in conjunction with Firefox's option `browser.sessionstore.restore_on_demand` set to `false` - This will cause session restore to immediately load all the tabs in the session, so it is all styled dark and beautiful as soon as you click it.

It's not a feature for every person or every PC, it's very demanding and some of my computers can't handle it for any session over a few tabs, but for the PCs with some CPU to spare, it really lets Dark Reader shine... or... Un-shine ;)

And now DR can work with a privacy userscript or addon such as https://addons.mozilla.org/en-US/firefox/addon/always-visible/ :)


But mostly, I just wanted to say, Thanks!
